### ISSUE TITLE
YAML switches or an equivalent functionality

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#958]
In GitLab by @mandresm on Feb 23, 2023, 08:28

Hello @dbeltrankyl and @kinow,

### Summary of the feature

I am facing a situation in which I need to choose between different dictionary keys depending on the HPC and then call that at the `jobs.yml`. This issue aims to open a discussion on how this composition of information can be achieve without adding more functionality to Autosubmit yaml parser, or whether on the contrary an implementation of switches is required. @mcastril, @dbeltrankyl, @kinow, @ jberlin, @sebastianbeyer 

### Describe your use case

In our `jobs.yaml` we have defined the following:

```yaml
JOBS:
    [ ... ]
    SIM:
        CUSTOM_DIRECTIVES: ["#SBATCH -p debug", "#SBATCH -N 1", "#SBATCH --ntasks-per-node=64", "#SBATCH --export=ALL"]
    [ ... ]
```

Let's say I now know that the number of nodes I'll be requesting depends on 1) my resolution type defined in the file `test.yml` and 2) the HPC in which I'm running the simulation.

One could write this in the `test.yml` in this way:
```yaml
CONFIGURATION:
    resolution:
        tco79:
            lumi:
                nnodes: 4
            juwels:
                nnodes: 8
```

Then the question is how do we select the correct `nnodes` in the `CUSTOM_DIRECTIVE` in `jobs.yaml`, if we would want to use place holders exclusively we would need a place holder inside another place holder (`%CONFIGURATION.resolution.%HPCARCH%.nnodes%`) which I think is not at all a good syntax:
```yaml
JOBS:
    [ ... ]
    SIM:
        CUSTOM_DIRECTIVES: ["#SBATCH -p debug", "#SBATCH -N %CONFIGURATION.resolution.%HPCARCH%.nnodes%", "#SBATCH --ntasks-per-node=64", "#SBATCH --export=ALL"]
    [ ... ]
```

The other option I can imagine is to have switches within yaml files, for example:
```yaml
CONFIGURATION:
    resolution:
        tco79:
            switch_HPCARCH:
                lumi:
                    nnodes: 4
                juwels:
                    nnodes: 8
```
In this way, the switch will be resolved and we could write in our custom directive `%CONFIGURATION.resolution.nnodes%`. We do this in the `esm_parser` of ESM-Tools, that, and list expansion, dictionary merging with hierarchical and switch considerations... and in my experience adding the switch functionality is like opening Pandora's Box. We figured it out after some years of trial error and users reporting issues.

So in summary, it seems to me that we will need this functionality for what we are trying to achieve, but if there is an alternative solution that is good enough I would rather not have switches in ClimateDT. Alternatively, we could use some parts of the `esm_parser` to enhance the yaml parsing here, but I would have to discuss this with the ESM-Tools team and with other people.

Let me know what you think, hopefully, I'm just overlooking something obvious.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 23, 2023, 09:08

Hello, 

If I understood well, in the end, what you want to achieve is to automatically set a configuration depending on the value of the HPCARCH? 

Maybe we can load a different custom_config depending on the value of the hpcarch.

I made some changes in !315(working already, but I want to do more tests),  allowing the custom_config directive to be defined inside another custom_config directive. Maybe this is enough. 

One could do now: 

* 1- Create an expid-minimal.yml ( or use a new expid -min parameter). 
* 2- In this expid-minimal.yml, Custom_config points to any proj configuration
* 3- This project configuration could have another custom_config pointing to another configuration.
* 4- $EXPID/conf configuration can now override project configuration.
* 5- We now have PRE and POST to define the loading order ( minimal.yml has an special case) 

I think the custom_config directive could be set to something like this (summarized to only the relevant right now).

expid-minimal.yml:

```
DEFAULT:
 HPCARCH: juwels
 CUSTOM_CONFIG: 
   PRE: "%ROOTDIR%/proj/git_project/conf_%HPCARCH%/" # proj data 
   POST: "/home/dbeltran/platforms.yml # user data
```

conf_juwels/test.yml could have: 

```
DEFAULT:
 CUSTOM_CONFIG: "%ROOTDIR%/proj/git_project/common/jobs.yml"
CONFIGURATION:
 resolution:
  tco79:
   nnodes: 8
```

conf_lumi/test.yml could have: 
```
DEFAULT:
 CUSTOM_CONFIG: "%ROOTDIR%/proj/git_project/common/jobs.yml"
CONFIGURATION:
   resolution:
    tco79:
     nnodes: 4

```

And common/jobs.yml could have:

```
JOBS:
    ...
    SIM:
        CUSTOM_DIRECTIVES: ["#SBATCH -p debug", "#SBATCH -N %CONFIGURATION.resolution.nnodes%", "#SBATCH --ntasks-per-node=64", "#SBATCH --export=ALL"]
    ...
```

Another alternative would be to load a different jobs.yml:

minimal.yml

```
DEFAULT:
 HPCARCH: juwels
 CUSTOM_CONFIG: 
   PRE: "%ROOTDIR%/proj/git_project/conf_%HPCARCH%/" # proj data 
   POST: "/home/dbeltran/platforms.yml # user data
```

conf_%HPCARCH%/test.yml

```
DEFAULT:
 CUSTOM_CONFIG: 
   PRE: "%ROOTDIR%/proj/git_project/common/jobs.yml" # load common jobs before override them
CONFIGURATION:
 ...
```


conf_juwels/jobs.yml could have: 

```
JOBS:
 SIM: 
   CUSTOM_DIRECTIVES: ["#SBATCH -p debug", "#SBATCH -N 8", "#SBATCH --ntasks-per-node=64", "#SBATCH --export=ALL"]
```


conf_lumi/jobs.yml could have: 
```
JOBS:
 SIM: 
  CUSTOM_DIRECTIVES: ["#SBATCH -p debug", "#SBATCH -N 4", "#SBATCH --ntasks-per-node=64", "#SBATCH --export=ALL"]

```


On offtopic, if we use the jobs exportable variables ( NODES,PROCESSORS, TASKS, THREADS, PLATFORM and PARTITION), we can avoid overloading all time the custom_directives I think the only one not defined for jobs is "#SBATCH --export=ALL"

**Comment by kinow** [comment]
In GitLab by @mandresm on Feb 23, 2023, 09:30

Considering that we are going to have yamls for resolution/ClimateDT_configuration, models, and simulation type, having one set of these yamls per HPC is not a solution I would favor. It's not modular and more difficult to maintain.

In my experience with working with modular configurations with yaml files, this is only an example, we are probably going to be hitting this wall many times.

> On offtopic, if we use the jobs exportable variables ( NODES,PROCESSORS, TASKS, THREADS, PLATFORM and PARTITION), we can avoid overloading all time the custom_directives I think the only one not defined for jobs is "#SBATCH --export=ALL"

We'll include your suggested changes in one of next commits.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 23, 2023, 17:19

That's what I meant Dani, and thanks for the confirmation.

As I see it and, according to your comments, this would be cost-effective.

**Comment by kinow** [comment]
In GitLab by @mandresm on Feb 24, 2023, 08:55

Thanks for your replies to both of you!

I think this solves the problem of the HPCs because Autosubmit, in a way, has already a switch implemented specifically for HPCs (in `PLATFORMS`, we have LUMI and JUWELS and then Autosubmit selects which dictionary to use). However, as far as I understand this solution (although very elegant) doesn't fully solve the problem because of the following reasons:

1. For the specific case of the number of nodes, if we want to provide a default number of nodes this is going to depend on: the platform, the grid and the model. With the approach from @mcastril we would need to keep different `tco79.yml` for different models. That would be okay if most of the parameters in that file would be different for the different models, but that might not be the case. This is a really silly example anyway, because in most of the cases I could imagine that the operator would want to define the number of nodes manually from the `expdef.yml` but I think we will find this situation in other cases, and that's why I'd like to continue discussing a strategy. That being said, taking inspiration from @mcastril idea, we could always keep general parameters in `conf/platforms.yml`, `conf/grids/tco79.yml` ... and have files for specific combinations in `conf/grids/ifs-fesom/tco79.yml` and have some smart little script in python that figures out which files it needs to load for the specific simulation you are trying to run. I think you guys are working already in something like that. I also know that @sebastianbeyer was playing with a script that does something similar. Effectively, what this script would do is to solve the selection outside of the yamls, so that you don't need internal choose. This would be very modular and robust. The disadvantage is that a simulation will be the result of composing many different files. That might get tricky for people that are new to the structure of the workflow and its files, for instance, when they try to modify some of the parameters they would need to look around for a while.

2. We are going to have apps in the same workflow, if we set the default in the `PLATFORM` this will work for models, but I could imagine that maybe APPs need their own specific nodes depending in the circumstances, so I would rather have this define in the `SIM` job, but then the platform switch of Autosubmit wouldn't solve the problem right?

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 24, 2023, 17:29

Hi Miguel, thanks for your answer,

> and have some smart little script in python that figures out which files it needs to load for the specific simulation you are trying to run

Couldn't you load those files always? I mean:

```
 CUSTOM_CONFIG: 
   PRE: "%ROOTDIR%/proj/conf/experiments/%EXP_TYPE%.yml", "%ROOTDIR%/proj/conf/platforms/%HPCARCH%.yml", "%ROOTDIR%/proj/grids/%GRID%.yml", "%ROOTDIR%/proj/models/%MODEL%/%GRID%.yml"
```

Leaving the files empty when there are no differences

Maybe I am minimizing the complexity as you spent more time considering the different possibilities.

> so I would rather have this defined in the `SIM` job, but then the platform switch of Autosubmit wouldn't solve the problem, right?

We could only set one default, which wouldn't work for the applications. 

Honestly, I don't think it will happen for many applications, as most of them should only use a few threads.

Ultimately, we could add: "%ROOTDIR%/proj/applications/mhm/%HPCARCH%.yml/" . It's not super-elegant, but if it only happens in one or two applications, maybe we can afford it.

I add @Lerriola to the conversation.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 2, 2023, 10:59

> Hey Miguel, I like that new CUSTOM_CONFIG syntax and the structure you write in your last comment! The other idea with the loaded yaml files themselves loading their own CUSTOM_CONFIG recursively seems way too convoluted to me.

Thanks, the benefit of the last structure is that we need no recursion. All model-grid-platform-dependant info can be written in the last directory: `"%ROOTDIR%/proj/models/%MODEL%/%GRID%.yml/"`

There you could have declarations for the different platforms:

```
- lumi:
  - nodes: 8
- mn5
  - nodes: 6
```

For every model-grid pair.

> Does the syntax of your last post work already with the variable expansion?

Dani has to confirm if the last changes in CUSTOM_CONF, regarding PRE/POST declaration, are already in `pip`

> However, I don't see how that solves the problem of the SLURM directives depending on multiple factors.

Explained above. I suggest having, for every model, one grid file declaring the recommended node settings per every platform. This can be achieved with just one directory depth. 

Of course, other model and grid dependant settings can be declared there, overloading the grid defaults.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 2, 2023, 11:03

Also, remind that if the operator wants to overload any parameter, they can do this in the local `.conf` file in `/conf`

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 3, 2023, 11:58

Refining my previous structure

**minimal-config.yml**
```
 RUN:
  - HPCARCH: MN4
  - EXP_TYPE: HISTORICAL
  - MODEL: IFS-NEMO
  - GRID: Tco1279-ORCA12
 CUSTOM_CONFIG: 
   PRE: "%ROOTDIR%/proj/conf/experiments/%EXP_TYPE%.yml", "%ROOTDIR%/proj/conf/platforms/%HPCARCH%.yml", "%ROOTDIR%/proj/grids/%GRID%.yml", "%ROOTDIR%/proj/models/%MODEL%/%GRID%.yml"
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 3, 2023, 12:02

Shouldn't it be:

```
DEFAULT:
  HPCARCH: MN4
  CUSTOM_CONFIG: 
   PRE: "%PROJDIR%/conf/experiments/%RUN.EXP_TYPE%.yml", "%PROJDIR%/conf/platforms/%DEFAULT.HPCARCH%.yml", "%PROJDIR%/grids/%RUN.GRID%.yml", "%PROJDIR%/models/%RUN.MODEL%/%RUN.GRID%.yml"


RUN:
  EXP_TYPE: HISTORICAL
  MODEL: IFS-NEMO
  GRID: Tco1279-ORCA12
```

I also replaced ROOTDIR per PROJDIR ( new parameter that points to proj/proj_destination)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 3, 2023, 12:03

or:

```
DEFAULT:
  HPCARCH: MN4
  CUSTOM_CONFIG: 
   PRE: 
    - "%PROJDIR%/conf/experiments/%RUN.EXP_TYPE%.yml"
    - "%PROJDIR%/conf/platforms/%DEFAULT.HPCARCH%.yml" 
    - "%PROJDIR%/grids/%RUN.GRID%.yml" 
    - "%PROJDIR%/models/%RUN.MODEL%/%RUN.GRID%.yml"


RUN:
  EXP_TYPE: HISTORICAL
  MODEL: IFS-NEMO
  GRID: Tco1279-ORCA12
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 3, 2023, 15:05

Hello @sebastianbeyer , 

Fixed some issues. I've tested the example you gave me in the matter most, and it is currently working with the v4.0.0b9 tag for your current structure. 


Generated a new expid with:

`autosubmit expid -min -repo https://earth.bsc.es/gitlab/digital-twins/de_340/workflow -b ifs-fesom-selectable-exp -d "ifs-fesom"`

minimal-yml modified to 

```
CONFIG:
  AUTOSUBMIT_VERSION: "4.0.0b"
  TOTALJOBS: 20
  MAXWAITINGJOBS: 20
DEFAULT:
  EXPID: "a06m"
  HPCARCH: "LUMI"
  #hint: use %PROJDIR% to point to the project folder (where the project is cloned)
  CUSTOM_CONFIG:
    PRE:
     - "%PROJDIR%/conf"
     - "%PROJDIR%/conf/simulation/%RUN.SIMULATION%.yml"
     - "%PROJDIR%/conf/configuration/%RUN.CONFIGURATION%.yml"
PROJECT:
  PROJECT_TYPE: "git"
  PROJECT_DESTINATION: "git_project"
GIT:
  PROJECT_ORIGIN: "https://earth.bsc.es/gitlab/digital-twins/de_340/workflow"
  PROJECT_BRANCH: "ifs-fesom-selectable-exp"
  PROJECT_COMMIT: ""
  PROJECT_SUBMODULES: ""
  FETCH_SINGLE_BRANCH: True
  
RUN:
  SIMULATION: test
  CONFIGURATION: test
```

The only thing I changed in the repo (locally) as it had the wrong syntax. 

is the %PROJDIR%/conf/jobs.yml

```
SIM:
  CUSTOM_DIRECTIVES: "[#SBATCH -p debug, #SBATCH -N 1, #SBATCH --ntasks-per-node=64, #SBATCH --export=ALL]"
```


Others things:

Finally added the minimal files to pip. When we switched to make a sdist it was no longer reading the files folder; thus, it was not uploading the minimal files. I added a MANIFEST.in instead of relying on the setup.py, and now it works.

* Added local-minimal.yml
* Updated docs, although it is not working right now in readthedocs. 
* Added -local flag to autosubmit expid -min , which will use a local skeleton instead of git

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 3, 2023, 15:08

Just to clarify,

$EXPID/conf/minimal.yml has the lowest priority. The keys on there, if used while the experiment is running, can be overwritten. I recommend putting RUN stuff in a different file. Per example

$expid/conf/minimal.yml
```
CONFIG:
  AUTOSUBMIT_VERSION: "4.0.0b"
  TOTALJOBS: 20
  MAXWAITINGJOBS: 20
DEFAULT:
  EXPID: "a06m"
  HPCARCH: "LUMI"
  #hint: use %PROJDIR% to point to the project folder (where the project is cloned)
  CUSTOM_CONFIG:
    PRE:
     - "%PROJDIR%/conf"
     - "%PROJDIR%/conf/simulation/%RUN.SIMULATION%.yml"
     - "%PROJDIR%/conf/configuration/%RUN.CONFIGURATION%.yml"
PROJECT:
  PROJECT_TYPE: "git"
  PROJECT_DESTINATION: "git_project"
GIT:
  PROJECT_ORIGIN: "https://earth.bsc.es/gitlab/digital-twins/de_340/workflow"
  PROJECT_BRANCH: "ifs-fesom-selectable-exp"
  PROJECT_COMMIT: ""
  PROJECT_SUBMODULES: ""
  FETCH_SINGLE_BRANCH: True
```

$EXPID/conf/run.yml # this file can only be overwritten by files in ( POST: in minimal.yml )

```
RUN:
  SIMULATION: test
  CONFIGURATION: test
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 3, 2023, 15:18

Side note: @mcastril , I can download the configuration and the templates. Which is enough to create refresh and inspect


But I have no access to git code.

```
git@gitlab.dkrz.de:
git@git.ecmwf.int: Permission denied (publickey).
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 5, 2023, 21:22

Thanks for the improvement @dbeltrankyl

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 5, 2023, 21:26

You need access to at least ECMWF bitbucket. 

Can you append a row for you here? https://earth.bsc.es/gitlab/digital-twins/de_340/project_management/-/issues/278#note_200011

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 5, 2023, 21:27

> Added -local flag to autosubmit expid -min , which will use a local skeleton instead of git

Dani, I am not sure of what the Git skeleton would be

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 5, 2023, 21:34

I see, we need a configuration file that works according to the rules defined in `CUSTOM-CONFIG`: `PRE`, local, `POST`.

So you suggest having the minimal configuration (the one used to load the CUSTOM_CONF) in `$EXPID/conf/minimal.yml` and loading the rest of the settings somewhere else. 

It's not always a problem that it's quite clear that `$EXPID/conf/minimal.yml` has the lowest priority. 

> $EXPID/conf/run.yml # this file can only be overwritten by files in ( POST: in minimal.yml )

Can you do that? Note that you are using `"%PROJDIR%/conf/simulation/%RUN.SIMULATION%.yml"` in `$expid/conf/minimal.yml` and this is declared in `$EXPID/conf/run.yml`. I would expect that it has to be declared in `$expid/conf/minimal.yml` as it has the lowest priority...

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 6, 2023, 09:01

Yes, you can do that. You can also put it in the minimal.yml. The only difference is that the minimal.yml only has priority while you're performing an autosubmit refresh ( or autosubmit create if there is no proj).

In summary, all the keys defined on minimal.yml are added to the final running configuration only in the case that the key is not defined elsewhere. 

### In refresh:

Minimal.yml is read, but it is only used for retrieve proj/projfiles, afterwards it is discarded.

However, the rest of configurarion in $expid/conf is read aswell

### In running( and rest of commands):

First, Pre keys of minimal.yml

Second, $expid/conf/*.yml(not minimal).

Third, post of $expid/conf/minimal.yml

And finally, minimal.yml keys are added only if the key doesnt exist in the unified configuration.

I hope now is more clear

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 6, 2023, 09:05

I would only put the minium neccesary to obtain the project data. That is what Autosubmit provides when using the -min flag

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 6, 2023, 09:49

mentioned in issue digital-twins/de_340/project_management#118

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 6, 2023, 11:02

To be more specific, what is the Git skeleton? Is something stored in the model project? I would like to understand better how this works, in contrast to using `local`

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 6, 2023, 11:10

Ah, it just generates a file with the local-specific variables or the git-specific variables:

GIT: 

`autosubmit expid -min -repo https://earth.bsc.es/gitlab/digital-twins/de_340/workflow -b ifs-fesom-selectable-exp -d "ifs-fesom"`

```
CONFIG:
  AUTOSUBMIT_VERSION: 4.0.0b
  TOTALJOBS: 20
  MAXWAITINGJOBS: 20
DEFAULT:
  EXPID: "$new_expid" 
  HPCARCH: "" # if not specified, it will be local
  #hint: use %PROJDIR% to point to the project folder (where the project is cloned)
  CUSTOM_CONFIG: "" # Can be filled with -conf
PROJECT:
  PROJECT_TYPE: "git"
  PROJECT_DESTINATION: "git_project"
GIT:
  PROJECT_ORIGIN: "https://earth.bsc.es/gitlab/digital-twins/de_340/workflow"
  PROJECT_BRANCH: "ifs-fesom-selectable-exp"
  PROJECT_COMMIT: ""
  PROJECT_SUBMODULES: ""
  FETCH_SINGLE_BRANCH: True
```

LOCAL: 

`autosubmit expid -min -local -repo /home/dbeltran/ifs-fesom-local -d "ifs-fesom" `


```
CONFIG:
  AUTOSUBMIT_VERSION: 4.0.0b
  TOTALJOBS: 20
  MAXWAITINGJOBS: 20
DEFAULT:
  EXPID: "$new_expid"
  HPCARCH: "local" # if not specified, it will be local
  #hint: use %PROJDIR% to point to the project folder (where the project is cloned)
  CUSTOM_CONFIG: "" # Can be filled with -conf
project:
  PROJECT_TYPE: "local"
  PROJECT_DESTINATION: "local_project"
local:
  PROJECT_PATH: '/home/dbeltran/ifs-fesom-local'
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 6, 2023, 11:11

Thanks @dbeltrankyl , I think it's clearer now.

So the difference is, as you say, that the parameters declared in `refresh` only have priority to take the project configurations.

> In running( and rest of commands):

However, I see something is missing here. You said that defining the parameters in another file would work the same, and in your workflow the local files are only processed after PRE. Could you refine the workflow taking into account also the rest of the files?

I understand that in this scenario all local files should be loaded first, before using PRE in minimal.yml

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 6, 2023, 11:13

Oh, I see, thanks!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 6, 2023, 11:30

> However, I see something is missing here. You said that defining the parameters in another file would work the same, and in your workflow the local files are only processed after PRE. Could you refine the workflow taking into account also the rest of the files?

All files under $expid/conf/*.yml are read before parsing the minimal.yml CUSTOM_CONFIG. So if any variable is used by the minimal.yml if it is inside $expid/conf/, Autosubmit will know it. 

However, when Autosubmit begins to read the PRE key, what it knows at that step is a dictionary with only the common variables {ROOTDIR,PROJDIR}.

So: 

* First it reads whole $EXPID/conf and substitutes  %placeholders% variables 

* In create(without proj)/refresh: reads {Minimal.yml , $expid/conf/} -> Downloads proj data -> resets config -> reads %ROOTDIR%, %PROJDIR% -> MINIMAL{PRE,with %% substituted} -> $expid/conf -> MINIMAL{POST, with %% substituted} -> reads minimal.yml(minimal does not overwrites only adds if key doesn't exists).  

* Rest of commands: only looks at minimal.yml custom_config parsed key -> reads %ROOTDIR%, %PROJDIR% -> MINIMAL{PRE, with %% substituted} -> $expid/conf -> MINIMAL{POST. with %% substituted} -> minimal.yml( minimal does not overwritten only adds if key doesn't exists) 


> I understand that in this scenario all local files should be loaded first, before using PRE in minimal.yml

Yes, that is what it is doing now.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 6, 2023, 11:55

Thanks @dbeltrankyl, that's better

So 

> The only difference is that the minimal.yml only has priority while performing an autosubmit refresh ( or autosubmit create if there is no proj).

It has priority, in terms that it's loaded first, but it would be overwritten anyway with anything loaded afterward.

So, it could be easily clarified by saying that the workflow can rely on `minimal.yml` to load external configurations the first time, but not later. And the indication should be to declare local, additional parameters in a specific file.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 6, 2023, 11:57

> So, it could be easily clarified by saying that the workflow can rely on `minimal.yml` to load external configurations the first time, but not later. And the indication should be to declare local, additional parameters in a specific file.

Agree

**Comment by kinow** [comment]
I will try to catch up on this thread later to learn more about the new configuration changes. Even though I code-reviewed it, looks like there are certain terms and peculiarities that I am still not familiar with.

@dbeltrankyl in case I find anything useful to document, is there a page in our docs you would suggest to add/update the information about the minimal configuration?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 6, 2023, 12:33

I've updated

https://autosubmit.readthedocs.io/en/master/userguide/set_and_share_the_configuration/index.html

I think that the latest version of this page it is still not being shown in the docs, and it needs a few modifications to make it clearer, taking into account Miguel latest comments

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 6, 2023, 12:56

Done, I thought I had already access due to the deployment of raps a long time ago, but no

**Comment by kinow** [comment]
In GitLab by @sebastianbeyer on Mar 7, 2023, 20:20

@dbeltrankyl Thank you! I was at the hackathon at ECMWF and could only test it today. I confirm that it works, great!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 10, 2023, 10:53

Hello @kinow , @mcastril 

What we should do with the old parameter: 

PROJECT_FILES ? 

This parameter copies a jobs/proj configuration files from the %PROJDIR% conf to the $expid/conf folder, whenever the user performs an `autosubmit refresh`

@charlio86 , reported to me that in the last tag, it doesn't work anymore. However, I see it deprecated because the users can now use the new CUSTOM_CONFIG. 

What do you think?

**Comment by kinow** [comment]
Found this `PROJECT_FILES` in one of our test files (not used/run) in the Autosubmit sources.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/b26104073a70913a704126d36ec5c0d7/image.png)

It does look like that's not useful any longer as the project/jobs configuration could be anywhere on the filesystem, and in one or multiple files/keys being loaded. +1 to deprecation & removal.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 12, 2023, 21:29

Yes, I agree that we can deprecate it in AS4.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 30, 2023, 09:34

I'll close this issue, 

Will continue https://earth.bsc.es/gitlab/es/autosubmit/-/issues/1004 for project_files deletion

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 12, 2023, 12:58

mentioned in issue digital-twins/de_340/workflow#10

**Comment by kinow** [comment]
In GitLab by @ainagaya on Apr 25, 2023, 17:12

mentioned in merge request digital-twins/de_340/workflow!18
